### DISCUSSION TITLE
Not possible to use an unmapped dataclass, which inherits from a mapped dataclass?

### DISCUSSION THREAD

**Discussion by HHoofs** [sqlalchemy/sqlalchemy#11607] [Usage Questions]
I have a small MWE, which uses imperative mapping. Within this example I first setup all aspects to create and map a ```Foo``` model and associated table. I furthermore add a single ```Foo``` instance to the table using a session. Afterwards I want to create an instance of the ```Bar``` model, which inherits from the ```Foo``` model, but is not (yet) mapped. Is it possible to get this working, without mapping the ```Bar``` first to the metadata.

```python
# SQLAlchemy==2.0.31
# Python 3.12.3 (macOS-14.5-arm64-arm-64bit)
from dataclasses import dataclass
from sqlalchemy import (
    Column,
    Integer,
    MetaData,
    String,
    Table,
    create_engine,
)
from sqlalchemy.orm import registry, Session


@dataclass
class Foo:
    """Basic dataclass"""
    bar: str


def create_foo_table(metadata: MetaData, engine) -> Table:
    """Create basic foo table to map Foo dataclass against"""
    foo_table = Table(
        'foo',
        metadata,
        Column('id', Integer, primary_key=True),
        Column('bar', String),
    )
    metadata.create_all(engine)  # create table
    return foo_table


@dataclass
class Bar(Foo):
    """
    Inherits all attributes from Foo (bar) and adds its
    own attributes (foo)

    :param bar: attribute from the the parent class
    :param foo: attribute initated by this class itself
    """
    foo: str


def create_bar_table(metadata: MetaData, engine) -> Table:
    """Create basic bar table to map Bar dataclass against"""
    bar_table = Table(
        'bar',
        metadata,
        Column('id', Integer, primary_key=True),
        Column('bar', String),
        Column('foo', String)
    )
    metadata.create_all(engine)  # create table
    return bar_table


if __name__ == '__main__':
    # Setup MWE
    mapper_registry = registry()
    engine = create_engine("sqlite://", future=True, echo=False)
    foo_table = create_foo_table(mapper_registry.metadata, engine)
    mapper_registry.map_imperatively(Foo, foo_table)

    # Add foo instance to the database
    with Session(engine) as foo_session:
        foo = Foo('bar')
        foo_session.add(foo)
        foo_session.commit()

    try:
        Bar('bar', 'foo')
    except AttributeError as e:
        print(e)  # 'AttributeError: 'Bar' object has no attribute '_sa_instance_state'

    # If Bar is also mapped everythin works as expected
    bar_table = create_bar_table(mapper_registry.metadata, engine)
    mapper_registry.map_imperatively(Bar, bar_table)  
    Bar('bar', 'foo')
```

**Comment by zzzeek** [comment]
a Python dataclass creates a constructor for each new class, and in this case `Bar` gets a new constructor that does not call super().__init__(), which means SQLAlchemy's augmentation to the constructor on `Foo` does not get called, leading to the `_sa_instance_state` attribute being missing.

I tried a few ways to illustrate how you can give an explicit `__init__` to `Bar`, but it's complicated as you have to fight against the private methodology of dataclasses and the private methodology of SQLAlchemy at the same time, and putting an explicit `__init__` on`Bar` defeats the purpose of it being a dataclass in the first place so I didn't go further in figuring that out.

if you want `Bar` to behave 99% like it's unmapped, you can make it a single-inheritance subclass of `Foo` without much issue, but I dont know how it's useful for `Bar` to be unmapped since `foo` wont get persisted and you can't load any `Bar` objects from the database without a polymorphic identity.

```
    mapper_registry.map_imperatively(Bar, None, inherits=Foo)
```

**Reply by HHoofs** [reply]
Thanks for the swift response. The problem is that in my (not so) minimal (not so) working example, I want to map the Bar class eventually, and I have no control whether it is already mapped or not...

**Reply by zzzeek** [reply]
you can control that if you use a metaclass that maps the class when it's created, or a base that includes `__init_subclass__()`.   that's after all what the declarative base does.

I guess you're writing an object framework of some kind.

long story short I can show you a dozen ways to get the correct initializer onto that class, mapped or not, but it requires a step that has to take place in some way that is not part of the `@dataclass` decorator, so you'd have to decide what kind of hook you'd be OK with.

**Reply by instanceofmel** [reply]
In my case I'm not using a `@dataclass` -- how can I get the correct initializer in this setup without mapping to `None`?

```python
import abc
from sqlalchemy import Column, ForeignKeyConstraint, String, Table, Uuid
from sqlalchemy.orm import registry


registry = registry()


class Employee(abc.ABC):
    def __init__(self, type, salary):
        self.type = type
        self.salary = salary

    @property
    def bonus(self):
        return self.salary * 0.01


class Engineer(Employee):
    def __init__(self):
        super().__init__("engineer", 100000)


employee_table = Table(
    "employee",
    registry.metadata,
    Column("id", Uuid(), primary_key=True),
    Column("type", String(16), nullable=False),
    # salary...
)


engineer_table = Table(
    "engineer",
    registry.metadata,
    Column("id", Uuid(), primary_key=True),
    ForeignKeyConstraint(["id"], ["employee.id"]),
)


employee_mapper = registry.map_imperatively(
    Employee, employee_table, polymorphic_on="type", polymorphic_abstract=True
)

registry.map_imperatively(
    Engineer, engineer_table, polymorphic_identity="engineer", inherits=employee_mapper
)


engineer = Engineer()  # Works perfectly


# When testing with ORM on ...
class MockEmployee(Employee):
    def __init__(self):
        super().__init__("mock", 4242)


# Line below doesn't work ... error: TypeError: MockEmployee.__init__() takes 1 positional argument but 3 were given
mock = MockEmployee()

# But works after mapping to None ...
registry.map_imperatively(MockEmployee, None, inherits=employee_mapper)
mock = MockEmployee()
```

Don't know if my case fits this discussion. Another error i get is `... object has no attribute '_sa_instance_state'` (for example when trying to add such an unmapped inheriting object to a list of a mapped class)

**Reply by zzzeek** [reply]
offlabel use but you can (maybe?) ignore those args in this case, basically there's a decorator around `__init__` that was designed to work with `Employee`, not `MockEmployee`.

overall though the Employee object has a lot of behaviors that the mapping gives it which you are relying upon.   so not sure what `MockEmployee` accomplishes

**Reply by zzzeek** [reply]
that is

```
class MockEmployee(Employee):
    def __init__(self, *args):
        super().__init__("mock", 4242)

```

**Reply by instanceofmel** [reply]
The model I provided here is just an example for reproduction, but my application does have a meaningful usecase in this constellation.

In case the salary has a default value in the original constructor, I get a different error (`MockEmployee.__init__() got multiple values for argument 'salary'`)

```python
import abc
from sqlalchemy import Column, ForeignKeyConstraint, String, Table, Uuid
from sqlalchemy.orm import registry


registry = registry()


class Employee(abc.ABC):
    def __init__(self, type, salary=None):
        self.type = type
        self.salary = salary

    @property
    def bonus(self):
        return self.salary * 0.01


class Engineer(Employee):
    def __init__(self):
        super().__init__("engineer", 100000)


employee_table = Table(
    "employee",
    registry.metadata,
    Column("id", Uuid(), primary_key=True),
    Column("type", String(16), nullable=False),
    # salary...
)


engineer_table = Table(
    "engineer",
    registry.metadata,
    Column("id", Uuid(), primary_key=True),
    ForeignKeyConstraint(["id"], ["employee.id"]),
)


employee_mapper = registry.map_imperatively(
    Employee, employee_table, polymorphic_on="type", polymorphic_abstract=True
)

registry.map_imperatively(
    Engineer, engineer_table, polymorphic_identity="engineer", inherits=employee_mapper
)


engineer = Engineer()  # Works perfectly


# When testing with ORM on ...
class MockEmployee(Employee):
    def __init__(self, salary):
        super().__init__("mock", salary)


# Line below doesn't work ... error: TypeError: MockEmployee.__init__() got multiple values for argument 'salary'
mock = MockEmployee(4242)

# But works after mapping to None ...
registry.map_imperatively(MockEmployee, None, inherits=employee_mapper)
mock = MockEmployee(4242)
```

Fixed it by something like

```python
class MockEmployee(Employee):
    def __init__(self, *args, **kwargs):
        super().__init__("mock", kwargs["salary"])

mock = MockEmployee(salary=4242)
```

I know its a super edge case, but adding `*args, **kwargs*` or mapping the class to `None` just to be able to instantiate an object (that inherits from a class thats mapped) feels weird.

Thanks for the tip with `*args` -- that got me on the right track. I'll go with the `*args, **kwargs*` workaround for now (y)

**Reply by CaselIT** [reply]
I don't understand why you need a subclass here.
Doesn't this work equally well?
```
def MockEmployee(salary):
  return Employee("mock", salary)

**Reply by instanceofmel** [reply]
Unfortunately no, since Employee is an ABC and cannot be instantiated

**Reply by CaselIT** [reply]
If it has no abstracmethods you should be able to instantiate it in any case

**Reply by instanceofmel** [reply]
Oh yeah, sorry, I forgot to add one in the example. In my real world scenario I have `abstractmethod`'s
